File: test/cs-stub/BaboonTests/KeyCodecHostLastWinsTests.cs
// PR-26.2 (M26) — KeyCodec Host last-wins concurrency contract regression test.
//
// Asserts FStr_KeyCodecHost.Register(impl) overwrites the previously registered
// impl (last-wins). C# already used a `volatile static` mutable singleton
// pre-PR-26.2; this test pins that behavior across future refactors.
//
// Generated symbols are produced by mdl :test-gen-regular-adt under
// target/test-regular/cs-stub/.
#nullable enable

using System.Collections.Generic;
using Baboon.Runtime.Shared;
using My.Ok.M19.Foreign;
using NUnit.Framework;

namespace ConversionsTest
{
    [TestFixture]
    public class KeyCodecHostLastWinsTests
    {
        private readonly BaboonCodecContext _ctx = BaboonCodecContext.Compact;

        private sealed class PrefixCodec : FStr_KeyCodec
        {
            private readonly string _tag;
            public PrefixCodec(string tag) { _tag = tag; }
            public string EncodeKey(string value) => _tag + ":" + value;
            public string DecodeKey(string s)
            {
                var pfx = _tag + ":";
                return s.StartsWith(pfx) ? s.Substring(pfx.Length) : s;
            }
        }

        private sealed class IdentityCodec : FStr_KeyCodec
        {
            public string EncodeKey(string value) => value;
            public string DecodeKey(string s) => s;
        }

        [Test]
        public void RegisterBAfterRegisterAObservesB()
        {
            var m = new Dictionary<ItemKey, string> { [new ItemKey("k")] = "v" };
            var original = new Holder(m);

            FStr_KeyCodecHost.Register(new PrefixCodec("A"));
            var encodedA = Holder_JsonCodec.Instance.Encode(_ctx, original).ToString();
            Assert.That(encodedA, Does.Contain("A:k"),
                "expected A: prefix in encoded wire form");

            FStr_KeyCodecHost.Register(new PrefixCodec("B"));
            var encodedB = Holder_JsonCodec.Instance.Encode(_ctx, original).ToString();
            Assert.That(encodedB, Does.Contain("B:k"),
                "PR-26.2 last-wins regression: expected B: prefix after re-register");
            Assert.That(encodedB, Does.Not.Contain("A:k"),
                "PR-26.2 last-wins regression: A: prefix still present after B re-register");

            // Restore identity-encoding default for any subsequent tests in this AppDomain.
            FStr_KeyCodecHost.Register(new IdentityCodec());
        }
    }
}

File: test/sw-stub/Tests/RuntimeTests/KeyCodecHostLastWinsTests.swift
// PR-26.2 (M26) — KeyCodec Host last-wins concurrency contract regression test.
//
// Asserts FStr_KeyCodecHost.register(impl) overwrites the previously registered
// impl (last-wins). Swift used a `nonisolated(unsafe) static var` mutable
// singleton pre-PR-26.2; this test pins that behavior across future refactors.
//
// Generated symbols are produced by mdl :test-gen-regular-adt under
// target/test-regular/sw-stub/.

import XCTest
import Foundation
@testable import BaboonRuntime
@testable import MyOkM19Foreign

final class KeyCodecHostLastWinsTests: XCTestCase {

    private let ctx = BaboonCodecContext.compact

    private struct PrefixCodec: FStr_KeyCodec {
        let tag: String
        func encodeKey(_ value: FStr) -> String { return "\(tag):\(value)" }
        func decodeKey(_ s: String) throws -> FStr {
            let pfx = "\(tag):"
            return s.hasPrefix(pfx) ? String(s.dropFirst(pfx.count)) : s
        }
    }

    private struct IdentityCodec: FStr_KeyCodec {
        func encodeKey(_ value: FStr) -> String { return value }
        func decodeKey(_ s: String) throws -> FStr { return s }
    }

    // PR-26.2-D01: restore identity impl after each test so the global
    // FStr_KeyCodecHost singleton does not leak a PrefixCodec into sibling
    // tests sharing the process. Runs even on assertion failure.
    override func tearDown() {
        FStr_KeyCodecHost.register(IdentityCodec())
        super.tearDown()
    }

    func testRegisterBAfterRegisterAObservesB() throws {
        let original = Holder(m: [ItemKey(v: "k"): "v"])

        FStr_KeyCodecHost.register(PrefixCodec(tag: "A"))
        let encodedA = Holder_JsonCodec.instance.encode(ctx, original)
        let bytesA = try JSONSerialization.data(withJSONObject: encodedA, options: [.sortedKeys])
        let stringA = String(data: bytesA, encoding: .utf8)!
        XCTAssertTrue(stringA.contains("A:k"),
                      "expected A: prefix in encoded wire form, got \(stringA)")

        FStr_KeyCodecHost.register(PrefixCodec(tag: "B"))
        let encodedB = Holder_JsonCodec.instance.encode(ctx, original)
        let bytesB = try JSONSerialization.data(withJSONObject: encodedB, options: [.sortedKeys])
        let stringB = String(data: bytesB, encoding: .utf8)!
        XCTAssertTrue(stringB.contains("B:k"),
                      "PR-26.2 last-wins regression: expected B: prefix after re-register, got \(stringB)")
        XCTAssertFalse(stringB.contains("A:k"),
                       "PR-26.2 last-wins regression: A: prefix still present after B re-register, got \(stringB)")
    }
}
